Client: SSL Expiry Error Message and Warnings Fix: #3866

ThePhisch opened the pull request:
Added functionality to warn the user of SSL certs that are either expired _or about to_ expire.
Some functions are derived from `cryptography.x509` which by default raises a DeprecationWarning in Python 3.6, these warnings have been disabled.

ThePhisch commented:
i made a new pull request so that my branchname actually fulfills the requirements in the contributing guide (as pointed out by @dchristidis )

rcarpa commented on .gitignore:
```diff
@@ -65,3 +65,6 @@ tools/test.file.1G
 /.autotest
+
+# VS Code
+.vscode/
```
Don't add this to repository-wide .gitignore. It is specific to your personal dev env, so the correct place for it is in ~/.gitignore on your machine.

ThePhisch replied:
true. fixed it

Commit:
Common/Client: Server SSL Expiry Error Message; Fix #3866

Added a new exception ServerSSLCertificateExpiredException to be called
by lib/rucio/client/baseclient.py when the SSL certificate provided by
the server is invalid.

Also raises a warning if the certificate provided will expire soon.
Whether a warning is raised and how many weeks must remain until expiry
for the warning to be raised is read from the config file. Since the
crytography module required for the certificate analysis is about to be
deprecated for python 3.6, it is imported via the EXTRA_MODULES way.

Commit:
Decorative changes (incl. type hinting of date_to_str function)

Commit:
Tests: Wrote tests for the Client Cert Handling; Fix #3866

The test does not test the validation of the SSL certificate or how the
age of the cert is extracted. It only tests whether the appropriate
errors and warnings are raised/logged:

* ServerSSLCertificateExpiredException upon finding that the SSL-cert is
  invalid
* logging that the SSL Cert will expire/has expired upon receiving a
  valid cert that is about to expire.

mlassnig commented on lib/rucio/client/baseclient.py:
```diff
@@ -432,8 +448,18 @@ def _send_request(self, url, headers=None, type_='GET', data=None, params=None,
+
+        # check if ssl expires soon and if client warning is requested
+        if ssl_expirydate < datetime.now() + timedelta(weeks=self.sslexpiry_warnlimit_weeks) and self.sslexpiry_warnclient:
+            self.logger.warning(f'SSL Certificate will expire in less than {self.sslexpiry_warnlimit_weeks} weeks on {date_to_str(ssl_expirydate)}.')
```
This should state which certificate (client or server one... :-)
